render/render.js

```diff
@@ -161,86 +161,212 @@ module.exports = function($window) {
 		else if (old == null) createNodes(parent, vnodes, 0, vnodes.length, hooks, nextSibling, ns)
-		else if (vnodes == null) removeNodes(old, 0, old.length, vnodes)
+		else if (vnodes == null) removeNodes(old, 0, old.length, vnodes, recyclingParent)
 		else {
-			if (old.length === vnodes.length) {
-				var isUnkeyed = false
-				for (var i = 0; i < vnodes.length; i++) {
-					if (vnodes[i] != null && old[i] != null) {
-						isUnkeyed = vnodes[i].key == null && old[i].key == null
-						break
-					}
+			var start = 0, commonLength = Math.min(old.length, vnodes.length), originalOldLength = old.length, hasPool = false, isUnkeyed = false
+			for(; start < commonLength; start++) {
+				if (old[start] != null && vnodes[start] != null) {
+					if (old[start].key == null && vnodes[start].key == null) isUnkeyed = true
+					break
 				}
-				if (isUnkeyed) {
-					for (var i = 0; i < old.length; i++) {
-						if (old[i] === vnodes[i]) continue
-						else if (old[i] == null && vnodes[i] != null) createNode(parent, vnodes[i], hooks, ns, getNextSibling(old, i + 1, nextSibling))
-						else if (vnodes[i] == null) removeNodes(old, i, i + 1, vnodes)
-						else updateNode(parent, old[i], vnodes[i], hooks, getNextSibling(old, i + 1, nextSibling), recycling, ns)
-					}
-					return
+			}
+			if (isUnkeyed && originalOldLength === vnodes.length) {
+				for (start = 0; start < originalOldLength; start++) {
+					if (old[start] === vnodes[start] && !recyclingParent || old[start] == null && vnodes[start] == null) continue
+					else if (old[start] == null) createNode(parent, vnodes[start], hooks, ns, getNextSibling(old, start + 1, originalOldLength, nextSibling))
+					else if (vnodes[start] == null) removeNodes(old, start, start + 1, vnodes, recyclingParent)
+					else updateNode(parent, old[start], vnodes[start], hooks, getNextSibling(old, start + 1, originalOldLength, nextSibling), recyclingParent, ns)
 				}
+				return
 			}
-			recycling = recycling || isRecyclable(old, vnodes)
-			if (recycling) {
-				var pool = old.pool
+
+			if (isRecyclable(old, vnodes)) {
+				hasPool = true
 				old = old.concat(old.pool)
 			}
 
-			var oldStart = 0, start = 0, oldEnd = old.length - 1, end = vnodes.length - 1, map
+			var oldStart = start = 0, oldEnd = old.length - 1, end = vnodes.length - 1, map, o, v, oFromPool
+
 			while (oldEnd >= oldStart && end >= start) {
-				var o = old[oldStart], v = vnodes[start]
-				if (o === v && !recycling) oldStart++, start++
-				else if (o == null) oldStart++
-				else if (v == null) start++
-				else if (o.key === v.key) {
-					var shouldRecycle = (pool != null && oldStart >= old.length - pool.length) || ((pool == null) && recycling)
+				o = old[oldStart]
+				v = vnodes[start]
+				oFromPool = hasPool && oldStart >= originalOldLength
+				if (o === v && !oFromPool && !recyclingParent || o == null && v == null) oldStart++, start++
+				else if (o == null) {
+					if (isUnkeyed || v.key == null) {
```

Comment:
The old `unkeyed` detection` may actually fail to detect keyedness if

`old === [null, {tag: "x"}]` and `vnodes === [{tag:'x'}, null]`, triggering #2003 once again.